app/models/potion.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+class Potion < ApplicationRecord
+  belongs_to :game
+  belongs_to :canonical_potion, optional: true, class_name: 'Canonical::Potion'
+
+  has_many :potions_alchemical_properties, dependent: :destroy, inverse_of: :potion
+  has_many :alchemical_properties, through: :potions_alchemical_properties
+
+  validates :name, presence: true
+  validates :unit_weight, numericality: { greater_than_or_equal_to: 0, allow_blank: true }
+
+  before_validation :set_canonical_potion
+
+  def canonical_models
+    return Canonical::Potion.where(id: canonical_potion_id) if canonical_potion.present?
+
+    matching = Canonical::Potion.where('name ILIKE ?', name)
+    matching = matching.where('magical_effects ILIKE ?', magical_effects) if magical_effects.present?
+    matching = matching.where(**attributes_to_match) if attributes_to_match.any?
+
+    return matching if matching.blank? || alchemical_properties.none?
+
+    matching
+      .joins(:canonical_potions_alchemical_properties)
+      .where(association_query)
+      .group('canonical_potions.id')
+      .having('COUNT(*) >= ?', potions_alchemical_properties.length)
+  end
+
+  alias_method :canonical_potions, :canonical_models
+
+  private
+
+  def attributes_to_match
+    { unit_weight: }.compact
+  end
+
+  def set_canonical_potion
+    return unless canonical_models.count == 1
+    return if canonical_potion.present?
```

Comment:
Is this necessary? If the `canonical_potion` is present then the `canonical_models` method will return an `ActiveRecord::Relation` containing only that model. Therefore, this method will already have returned by the time this line runs.